composable-tasks/src/compose_tasks.rs

```diff
@@ -161,6 +166,102 @@ pub trait TaskCall:
         EgldOrEsdtTokenPayment::from(payment_out)
     }
 
+    fn smart_swap(
+        &self,
+        payment_for_current_task: EgldOrEsdtTokenPayment,
+        payments_to_return: &mut PaymentsVec<Self::Api>,
+        args: &mut ManagedVec<ManagedBuffer>,
+    ) -> EgldOrEsdtTokenPayment {
+        require!(
+            !payment_for_current_task.token_identifier.is_egld(),
+            "EGLD can't be swapped!"
+        );
+        let mut payment_in = payment_for_current_task.unwrap_esdt();
+
+        let args_cloned = args.clone();
+        let mut aggregated_payment_out =
+            self.get_empty_payment_out_from_smart_swap_args(args_cloned);
+
+        let mut args_iter = args.into_iter();
+
+        loop {
+            let routes_no = match args_iter.next() {
+                Some(count) => count.parse_as_u64().unwrap(),
+                None => break,
+            };
+
+            // take the input amount for the swap
+            let amount_in = match args_iter.next() {
+                Some(amount) => BigUint::from(amount),
+                None => break,
+            };
+
+            payment_in.amount = amount_in;
+
+            // take args for a router_swap
+            let router_args: ManagedVec<ManagedBuffer> = args_iter
+                .by_ref()
+                .take(ROUTER_SWAP_ARGS_LEN * routes_no as usize)
```

Comment:
Why not `take()` directly here? Why `as_ref`? I believe that creates a separate reference, though I'm a bit unfamiliar with how managed types actually work.